packages/core/utils/src/BitSet.js

```diff
@@ -0,0 +1,121 @@
+// @flow strict-local
+import nullthrows from 'nullthrows';
+
+// As our current version of flow doesn't support BigInt's, these values/types
+// have been hoisted to keep the flow errors to a minimum. This can be removed
+// if we upgrade to a flow version that supports BigInt's
+// $FlowFixMe
+type TmpBigInt = bigint;
+// $FlowFixMe
+const BIGINT_ZERO = 0n;
+// $FlowFixMe
+const BIGINT_ONE = 1n;
+// $FlowFixMe
+let numberToBigInt = (v: number): TmpBigInt => BigInt(v);
+
+let bitUnion = (a: TmpBigInt, b: TmpBigInt): TmpBigInt => a | b;
+
+export class BitSet<Item> {
+  _value: TmpBigInt;
+  _lookup: Map<Item, TmpBigInt>;
+  _items: Array<Item>;
+
+  constructor({
+    initial,
+    items,
+    lookup,
+  }: {|
+    items: Array<Item>,
+    lookup: Map<Item, number>,
+    initial?: BitSet<Item> | TmpBigInt,
+  |}) {
+    if (initial instanceof BitSet) {
+      this._value = initial?._value;
+    } else if (initial) {
+      this._value = initial;
+    } else {
+      this._value = BIGINT_ZERO;
+    }
+
+    this._items = items;
+    this._lookup = lookup;
+  }
+
+  static from(items: Array<Item>): BitSet<Item> {
+    let lookup: Map<Item, TmpBigInt> = new Map();
+    for (let i = 0; i < items.length; i++) {
+      lookup.set(items[i], numberToBigInt(i));
+    }
+
+    return new BitSet({items, lookup});
+  }
+
+  static union(a: BitSet<Item>, b: BitSet<Item>): BitSet<Item> {
+    return new BitSet({
+      initial: bitUnion(a._value, b._value),
+      lookup: a._lookup,
+      items: a._items,
+    });
+  }
+
+  #getIndex(item: Item) {
+    return nullthrows(this._lookup.get(item), 'Item is missing from BitSet');
+  }
+
+  add(item: Item) {
+    this._value |= BIGINT_ONE << this.#getIndex(item);
+  }
+
+  delete(item: Item) {
+    this._value &= ~(BIGINT_ONE << this.#getIndex(item));
+  }
+
+  has(item: Item): boolean {
+    return Boolean(this._value & (BIGINT_ONE << this.#getIndex(item)));
+  }
+
+  intersect(v: BitSet<Item>) {
+    this._value = this._value & v._value;
+  }
+
+  union(v: BitSet<Item>) {
+    this._value = bitUnion(this._value, v._value);
+  }
+
+  clear() {
+    this._value = BIGINT_ZERO;
+  }
+
+  cloneEmpty(): BitSet<Item> {
+    return new BitSet({
+      lookup: this._lookup,
+      items: this._items,
+    });
+  }
+
+  clone(): BitSet<Item> {
+    return new BitSet({
+      lookup: this._lookup,
+      items: this._items,
+      initial: this._value,
+    });
+  }
+
+  values(): Array<Item> {
+    let values = [];
+    let tmpValue = this._value;
+    let i;
+
+    while (tmpValue > BIGINT_ZERO) {
+      // Get last set bit
+      i = tmpValue.toString(2).length - 1;
+
+      values.push(this._items[i]);
+
+      // Unset last set bit
+      tmpValue &= ~(BIGINT_ONE << numberToBigInt(i));
+    }
+
```

Comment:
You could do this in reverse order without the (probably expensive) string conversion on each iteration.

```js
let v = this._value;
for (let i = 0; v != 0n; i++, v >>= 1n) {
  if (b & 1n) values.push(this._items[i]);
}
```

Reply:
I originally went with this approach, but the trade-off is you end up iterating from `0..largestIndex` times. The current approach only iterates once per item in the set. This trade-off works in favour of large lists of potential values but small Set sizes. The string length hack is only needed because `Math.log2` isn't implemented for BigInts. 

We could potentially make a hybrid approach where if `this._value.toString(2).length < SOME_THRESHOLD` we use standard iteration?

Reply:
Did you measure which is faster? My guess was that bit shifting that many times would be faster that converting to a string since it doesn't require any memory allocation.

Reply:
Yeah I did. It's tricky as the length approach is faster, but only when the set sizes are really low. However when testing internal projects this trade-off seems to pay off. My project has around 68000 assets but averages about 50 internalised assets. 

This is modelling that in my benchmarks
```bash
values (string.length)
3.67ms avg | 739.17ms total | 200 runs

values (reverse loop)
36.22ms avg | 7243.16ms total | 200 runs
```

Reply:
Ok, nice. Might be good to add a comment to the code explaining that!

Reply:
Will do 👍

Reply:
Done